File: ccesim/simulator/recordset.py
"""
SimulatedRecordSet: the integration layer that orchestrates thermal, power,
and event models into schema-compatible record batches.
"""

import random
import datetime as dt
from dataclasses import dataclass, field
from typing import List, Optional

from ccesim.simulator.config import SimulationConfig
from ccesim.simulator.thermal import ThermalModel, ThermalState, AmbientModel, DoorEvent
from ccesim.simulator.power import MainsPowerModel, SolarPowerModel, PowerState
from ccesim.simulator.events import DoorEventGenerator, FaultInjector, AlarmGenerator
from ccesim.schemas import RtmdRecord, EmsRecordMains, EmsRecordSolar


@dataclass
class SimulatorState:
    """Persistent state carried between successive generate() calls.

    Allows a BaseRtmDevice to call generate() repeatedly (e.g., every
    upload interval) with continuity in TVC, battery SOC, etc.
    """
    tvc: float = 5.0
    compressor_on: bool = False
    battery_soc: float = 0.8
    cumulative_powered_s: float = 0.0
    rng_state: Optional[tuple] = None
    icebank_soc: float = 1.0  # Icebank state of charge (0.0–1.0)
    tvc_contents: Optional[float] = None  # Bulk contents temp (two-node model)
    # Internal state for sub-models (not user-facing)
    _power_in_outage: bool = False
    _power_outage_end: Optional[dt.datetime] = None
    _alarm_last_power_loss: Optional[dt.datetime] = None
    _alarm_power_was_available: bool = True
    _alarm_heat_excursion_start: Optional[dt.datetime] = None
    _alarm_frze_excursion_start: Optional[dt.datetime] = None
    _alarm_continuous_door_start: Optional[dt.datetime] = None
    _alarm_door_open_at_prev_end: bool = False


class SimulatedRecordSet:
    """A batch of simulated CCE records with conversion methods.

    Provides the same interface as the old RecordSet from ccedata.py:
      - .records: list of dicts
      - .to_rtmd() -> List[RtmdRecord]
      - .to_ems(powersource) -> List[EmsRecordMains|EmsRecordSolar]
      - .state: SimulatorState for continuity
    """

    def __init__(self, records: List[dict], state: SimulatorState, power_type: str):
        self.records = records
        self.state = state
        self._power_type = power_type

    @classmethod
    def generate(
        cls,
        config: SimulationConfig,
        batch_size: int,
        start_time: dt.datetime,
        interval: int = 900,
        state: Optional[SimulatorState] = None,
    ) -> 'SimulatedRecordSet':
        """Run the simulation and produce a batch of records.

        Args:
            config: Full simulation configuration.
            batch_size: Number of records to generate.
            start_time: Timestamp of the first record.
            interval: Seconds between records.
            state: Previous state for continuity. None = fresh start.

        Returns:
            SimulatedRecordSet with records and updated state.
        """
        # Initialize RNG
        if state is not None and state.rng_state is not None:
            rng = random.Random()
            rng.setstate(state.rng_state)
        elif config.random_seed is not None:
            rng = random.Random(config.random_seed)
        else:
            rng = random.Random()

        # Initialize state
        if state is None:
            state = SimulatorState(
                tvc=config.thermal.initial_tvc,
                battery_soc=config.power.battery_initial_soc,
                icebank_soc=config.thermal.icebank_initial_soc,
            )

        # Build sub-models
        thermal_model = ThermalModel(config.thermal)
        ambient_model = AmbientModel(config.ambient, rng)
        door_gen = DoorEventGenerator(config.events, rng)
        fault_injector = FaultInjector(config.fault, start_time)
        alarm_gen = AlarmGenerator(rng)

        # Restore alarm generator state
        alarm_gen._last_power_loss = state._alarm_last_power_loss
        alarm_gen._power_was_available = state._alarm_power_was_available
        alarm_gen._heat_excursion_start = state._alarm_heat_excursion_start
        alarm_gen._frze_excursion_start = state._alarm_frze_excursion_start
        alarm_gen._continuous_door_start = state._alarm_continuous_door_start
        alarm_gen._door_open_at_prev_end = state._alarm_door_open_at_prev_end

        # Build power model
        power_state = PowerState(
            cumulative_powered_s=state.cumulative_powered_s,
            battery_soc=state.battery_soc,
            in_outage=state._power_in_outage,
            outage_end=state._power_outage_end,
        )

        is_solar = config.power.power_type == "solar"
        if is_solar:
            power_model = SolarPowerModel(config.power, rng)
        else:
            power_model = MainsPowerModel(config.power, rng)

        # Thermal state
        thermal_state = ThermalState(
            tvc=state.tvc,
            compressor_on=state.compressor_on,
            icebank_soc=state.icebank_soc,
            tvc_contents=state.tvc_contents,
        )

        records = []

        for i in range(batch_size):
            timestamp = start_time + dt.timedelta(seconds=i * interval)

            # 1. Ambient temperature
            tamb = ambient_model.get_tamb(timestamp)

            # 2. Door events
            door_events = door_gen.get_door_events(timestamp, interval)

            # 3. Fault effects
            fault_effects = fault_injector.get_fault_effects(timestamp)

            # 4. Power availability
            if fault_effects.power_available_override is not None:
                power_available = fault_effects.power_available_override
            elif is_solar:
                power_available = power_model.is_power_available(power_state, timestamp)
            else:
                power_available = power_model.is_power_available(power_state)

            # 5. Compressor availability (power + fault)
            compressor_available = power_available and fault_effects.compressor_available

            # 6. Apply stuck door fault
            if fault_effects.door_forced_open:
                door_events = [DoorEvent(start_offset_s=0, duration_s=float(interval))]

            # 7. Compute Q_compressor override for refrigerant leak
            q_override = None
            if fault_effects.q_compressor_multiplier < 1.0:
                q_override = config.thermal.Q_compressor * fault_effects.q_compressor_multiplier

            # 8. Thermal simulation
            thermal_state, thermal_record = thermal_model.simulate_interval(
                state=thermal_state,
                tamb=tamb,
                interval_s=float(interval),
                compressor_available=compressor_available,
                door_events=door_events,
                q_compressor_override=q_override,
            )

            # 9. Power readings
            power_state, power_record = power_model.simulate_interval(
                state=power_state,
                timestamp=timestamp,
                interval_s=float(interval),
                compressor_runtime_s=float(thermal_record['CMPR']),
                power_available_override=fault_effects.power_available_override,
            )

            # 10. Alarms
            alarm_record = alarm_gen.derive_alarms(
                tvc=thermal_state.tvc,
                power_available=power_available,
                timestamp=timestamp,
                door_events=door_events,
                interval_s=float(interval),
            )

            # 11. Battery/BLOG for mains. A mains EMD/logger keeps its backup
            # battery topped up, so days-of-life-remaining sits near full with
            # small variation (BLOG max 9999 integer days per Annex; BEMD max 9999.9).
            if not is_solar:
                full_days = config.power.blog_full_days
                blog = round(max(0.0, min(9999, rng.gauss(full_days, full_days * 0.02))), 1)
                power_record['BLOG'] = blog
                power_record['BEMD'] = blog

            # Assemble the record
            record = {'ABST': timestamp}
            record.update(thermal_record)
            record.update(power_record)
            record.update(alarm_record)

            records.append(record)

        # Save state for next call
        new_state = SimulatorState(
            tvc=thermal_state.tvc,
            compressor_on=thermal_state.compressor_on,
            battery_soc=power_state.battery_soc,
            cumulative_powered_s=power_state.cumulative_powered_s,
            rng_state=rng.getstate(),
            icebank_soc=thermal_state.icebank_soc,
            tvc_contents=thermal_state.tvc_contents,
            _power_in_outage=power_state.in_outage,
            _power_outage_end=power_state.outage_end,
            _alarm_last_power_loss=alarm_gen._last_power_loss,
            _alarm_power_was_available=alarm_gen._power_was_available,
            _alarm_heat_excursion_start=alarm_gen._heat_excursion_start,
            _alarm_frze_excursion_start=alarm_gen._frze_excursion_start,
            _alarm_continuous_door_start=alarm_gen._continuous_door_start,
            _alarm_door_open_at_prev_end=alarm_gen._door_open_at_prev_end,
        )

        return cls(records, new_state, config.power.power_type)

    def to_rtmd(self) -> List[RtmdRecord]:
        """Convert records to RTMD format (subset of fields)."""
        rtmd_fields = ['ABST', 'BEMD', 'TVC', 'TAMB', 'ALRM', 'EERR']
        filtered = [{k: r[k] for k in rtmd_fields if k in r} for r in self.records]
        return [RtmdRecord(**rec) for rec in filtered]

    def to_ems(self, powersource: str = None) -> List[EmsRecordMains | EmsRecordSolar]:
        """Convert records to EMS format.

        Args:
            powersource: 'mains' or 'solar'. If None, inferred from config.
        """
        if powersource is None:
            powersource = self._power_type

        base_fields = [
            'ABST', 'ALRM', 'BEMD', 'BLOG', 'CMPR', 'DORV',
            'HAMB', 'HOLD', 'EERR', 'LERR', 'TAMB', 'TCON', 'TVC',
        ]

        if powersource == 'solar':
            extra_fields = ['DCCD', 'DCSV']
            schema = EmsRecordSolar
        else:
            extra_fields = ['ACCD', 'ACSV', 'SVA']
            schema = EmsRecordMains

        all_fields = base_fields + extra_fields
        result = []
        for r in self.records:
            # EERR (EMD error codes) and LERR (logger error codes) are
            # distinct fields; both are carried through independently.
            filtered = {k: r[k] for k in all_fields if k in r}
            result.append(schema(**filtered))

        return result

File: js/src/recordset.js
/**
 * SimulatedRecordSet: the integration layer that orchestrates thermal, power,
 * and event models into schema-compatible record batches.
 *
 * Port of ccesim/simulator/recordset.py
 */

import { SeededRandom } from "./random.js";
import { ThermalModel, ThermalState, AmbientModel, DoorEvent as ThermalDoorEvent } from "./thermal.js";
import { MainsPowerModel, SolarPowerModel, PowerState } from "./power.js";
import { DoorEventGenerator, FaultInjector, AlarmGenerator, DoorEvent } from "./events.js";
import { RtmdRecord, EmsRecordMains, EmsRecordSolar, formatEmsDateTime } from "./schemas.js";

/**
 * Persistent state carried between successive generate() calls.
 *
 * Allows a device to call generate() repeatedly with continuity
 * in TVC, battery SOC, etc.
 */
export class SimulatorState {
  constructor({
    tvc = 5.0,
    compressor_on = false,
    battery_soc = 0.8,
    cumulative_powered_s = 0.0,
    rng_state = null,
    icebank_soc = 1.0,
    tvc_contents = null,
    // Internal sub-model state
    _power_in_outage = false,
    _power_outage_end = null,
    _alarm_last_power_loss = null,
    _alarm_power_was_available = true,
    _alarm_heat_excursion_start = null,
    _alarm_frze_excursion_start = null,
    _alarm_continuous_door_start = null,
    _alarm_door_open_at_prev_end = false,
  } = {}) {
    this.tvc = tvc;
    this.compressor_on = compressor_on;
    this.battery_soc = battery_soc;
    this.cumulative_powered_s = cumulative_powered_s;
    this.rng_state = rng_state;
    this.icebank_soc = icebank_soc;
    this.tvc_contents = tvc_contents;
    this._power_in_outage = _power_in_outage;
    this._power_outage_end = _power_outage_end;
    this._alarm_last_power_loss = _alarm_last_power_loss;
    this._alarm_power_was_available = _alarm_power_was_available;
    this._alarm_heat_excursion_start = _alarm_heat_excursion_start;
    this._alarm_frze_excursion_start = _alarm_frze_excursion_start;
    this._alarm_continuous_door_start = _alarm_continuous_door_start;
    this._alarm_door_open_at_prev_end = _alarm_door_open_at_prev_end;
  }
}

/**
 * Convert events.js DoorEvent (start_offset_s, duration_s) to thermal.js
 * DoorEvent (startOffsetS, durationS) for the thermal model.
 * @param {Array} events - Array of {start_offset_s, duration_s} objects
 * @returns {Array} Array of ThermalDoorEvent objects
 */
function toThermalDoorEvents(events) {
  return events.map(e => new ThermalDoorEvent(e.start_offset_s, e.duration_s));
}

/**
 * A batch of simulated CCE records with conversion methods.
 */
export class SimulatedRecordSet {
  /**
   * @param {Array<object>} records - Array of record dicts.
   * @param {SimulatorState} state - State for continuity.
   * @param {string} powerType - "mains" or "solar".
   */
  constructor(records, state, powerType) {
    this.records = records;
    this.state = state;
    this._powerType = powerType;
  }

  /**
   * Run the simulation and produce a batch of records.
   *
   * @param {import('./config.js').SimulationConfig} config
   * @param {number} batchSize - Number of records to generate.
   * @param {Date} startTime - Timestamp of the first record.
   * @param {number} [interval=900] - Seconds between records.
   * @param {SimulatorState|null} [state=null] - Previous state for continuity.
   * @returns {SimulatedRecordSet}
   */
  static generate(config, batchSize, startTime, interval = 900, state = null) {
    // Initialize RNG
    let rng;
    if (state !== null && state.rng_state !== null) {
      rng = new SeededRandom(0);
      rng.setState(state.rng_state);
    } else if (config.random_seed !== null) {
      rng = new SeededRandom(config.random_seed);
    } else {
      rng = new SeededRandom(Date.now());
    }

    // Initialize state
    if (state === null) {
      state = new SimulatorState({
        tvc: config.thermal.initial_tvc,
        battery_soc: config.power.battery_initial_soc,
        icebank_soc: config.thermal.icebank_initial_soc,
      });
    }

    // Build sub-models
    const thermalModel = new ThermalModel(config.thermal);
    const ambientModel = new AmbientModel(config.ambient, rng);
    const doorGen = new DoorEventGenerator(config.events, rng);
    const faultInjector = new FaultInjector(config.fault, startTime);
    const alarmGen = new AlarmGenerator(rng);

    // Restore alarm generator state
    alarmGen._lastPowerLoss = state._alarm_last_power_loss;
    alarmGen._powerWasAvailable = state._alarm_power_was_available;
    alarmGen._heatExcursionStart = state._alarm_heat_excursion_start;
    alarmGen._frzeExcursionStart = state._alarm_frze_excursion_start;
    alarmGen._continuousDoorStart = state._alarm_continuous_door_start;
    alarmGen._doorOpenAtPrevEnd = state._alarm_door_open_at_prev_end;

    // Build power model
    const powerState = new PowerState({
      cumulative_powered_s: state.cumulative_powered_s,
      battery_soc: state.battery_soc,
      in_outage: state._power_in_outage,
      outage_end: state._power_outage_end,
    });

    const isSolar = config.power.power_type === "solar";
    const powerModel = isSolar
      ? new SolarPowerModel(config.power, rng)
      : new MainsPowerModel(config.power, rng);

    // Thermal state
    let thermalState = new ThermalState({
      tvc: state.tvc,
      compressorOn: state.compressor_on,
      icebankSoc: state.icebank_soc,
      tvcContents: state.tvc_contents,
    });

    const records = [];

    for (let i = 0; i < batchSize; i++) {
      const timestamp = new Date(startTime.getTime() + i * interval * 1000);

      // 1. Ambient temperature
      const tamb = ambientModel.getTamb(timestamp);

      // 2. Door events (events.js format: start_offset_s, duration_s)
      let doorEvents = doorGen.getDoorEvents(timestamp, interval);

      // 3. Fault effects
      const faultEffects = faultInjector.getFaultEffects(timestamp);

      // 4. Power availability
      let powerAvailable;
      if (faultEffects.power_available_override !== null) {
        powerAvailable = faultEffects.power_available_override;
      } else if (isSolar) {
        powerAvailable = powerModel.is_power_available(powerState, timestamp);
      } else {
        powerAvailable = powerModel.is_power_available(powerState);
      }

      // 5. Compressor availability (power + fault)
      const compressorAvailable = powerAvailable && faultEffects.compressor_available;

      // 6. Apply stuck door fault
      if (faultEffects.door_forced_open) {
        doorEvents = [DoorEvent(0, interval)];
      }

      // 7. Compute Q_compressor override for refrigerant leak
      let qOverride = null;
      if (faultEffects.q_compressor_multiplier < 1.0) {
        qOverride = config.thermal.Q_compressor * faultEffects.q_compressor_multiplier;
      }

      // 8. Thermal simulation (convert door events to thermal format)
      const thermalDoorEvents = toThermalDoorEvents(doorEvents);
      const [newThermalState, thermalRecord] = thermalModel.simulateInterval(
        thermalState,
        tamb,
        interval,
        compressorAvailable,
        thermalDoorEvents,
        qOverride,
      );
      thermalState = newThermalState;

      // 9. Power readings
      const [, powerRecord] = powerModel.simulate_interval(
        powerState,
        timestamp,
        interval,
        thermalRecord.CMPR,
        faultEffects.power_available_override,
      );

      // 10. Alarms (uses events.js door event format)
      const alarmRecord = alarmGen.deriveAlarms({
        tvc: thermalState.tvc,
        power_available: powerAvailable,
        timestamp,
        door_events: doorEvents,
        interval_s: interval,
      });

      // 11. Battery/BLOG for mains. A mains EMD/logger keeps its backup battery
      // topped up, so days-of-life-remaining sits near full with small variation
      // (BLOG max 9999 integer days per Annex; BEMD max 9999.9).
      if (!isSolar) {
        const fullDays = config.power.blog_full_days;
        const blog =
          Math.round(
            Math.max(0.0, Math.min(9999, rng.gauss(fullDays, fullDays * 0.02))) * 10
          ) / 10;
        powerRecord.BLOG = blog;
        powerRecord.BEMD = blog;
      }

      // Assemble the record
      const record = { ABST: timestamp };
      Object.assign(record, thermalRecord);
      Object.assign(record, powerRecord);
      Object.assign(record, alarmRecord);

      records.push(record);
    }

    // Save state for next call
    const newState = new SimulatorState({
      tvc: thermalState.tvc,
      compressor_on: thermalState.compressorOn,
      battery_soc: powerState.battery_soc,
      cumulative_powered_s: powerState.cumulative_powered_s,
      rng_state: rng.getState(),
      icebank_soc: thermalState.icebankSoc,
      tvc_contents: thermalState.tvcContents,
      _power_in_outage: powerState.in_outage,
      _power_outage_end: powerState.outage_end,
      _alarm_last_power_loss: alarmGen._lastPowerLoss,
      _alarm_power_was_available: alarmGen._powerWasAvailable,
      _alarm_heat_excursion_start: alarmGen._heatExcursionStart,
      _alarm_frze_excursion_start: alarmGen._frzeExcursionStart,
      _alarm_continuous_door_start: alarmGen._continuousDoorStart,
      _alarm_door_open_at_prev_end: alarmGen._doorOpenAtPrevEnd,
    });

    return new SimulatedRecordSet(records, newState, config.power.power_type);
  }

  /**
   * Convert records to RTMD format (subset of fields).
   * @returns {RtmdRecord[]}
   */
  toRtmd() {
    const rtmdFields = ["ABST", "BEMD", "TVC", "TAMB", "ALRM", "EERR"];
    return this.records.map(r => {
      const filtered = {};
      for (const k of rtmdFields) {
        if (k in r && r[k] !== undefined) {
          filtered[k] = r[k];
        }
      }
      return new RtmdRecord(filtered);
    });
  }

  /**
   * Convert records to EMS format.
   * @param {string|null} [powersource=null] - "mains" or "solar". If null, inferred from config.
   * @returns {Array<EmsRecordMains|EmsRecordSolar>}
   */
  toEms(powersource = null) {
    if (powersource === null) {
      powersource = this._powerType;
    }

    const baseFields = [
      "ABST", "ALRM", "BEMD", "BLOG", "CMPR", "DORV",
      "HAMB", "HOLD", "EERR", "LERR", "TAMB", "TCON", "TVC",
    ];

    let extraFields, Schema;
    if (powersource === "solar") {
      extraFields = ["DCCD", "DCSV"];
      Schema = EmsRecordSolar;
    } else {
      extraFields = ["ACCD", "ACSV", "SVA"];
      Schema = EmsRecordMains;
    }

    const allFields = [...baseFields, ...extraFields];

    return this.records.map(r => {
      // EERR (EMD error codes) and LERR (logger error codes) are distinct
      // fields; both are carried through independently.
      const filtered = {};
      for (const k of allFields) {
        if (k in r && r[k] !== undefined) {
          filtered[k] = r[k];
        }
      }
      return new Schema(filtered);
    });
  }
}
